File: JS-Solutions/Problem-14.js
// Which starting number, under one million, produces the longest Collatz chain?

collatzLength = (num) => {
  let len = 1;
  while (num > 1) {
    if (num % 2 == 0) {
      num /= 2;
    } else {
      num = num*3+1;
    }
    len += 1
  }
  return len
}

maxCollatz = (limit) => {
  let answer = 0;
  let max_len = 0;
  for (i=3;i<limit;i+=2) {
    let len = collatzLength(i);
    if (len > max_len) {
      answer = i;
      max_len = len;
    }
  }
  return [answer, max_len];
}

console.log(maxCollatz(1000000));

File: Swift-Solutions/Problem-14.swift
// Which starting number, under one million, produces the longest Collatz chain?

func collatzLength(of num: Int) -> Int {
  var length: Int = 1
  var tracker: Int = num
  while tracker > 1 {
    if tracker % 2 == 0 {
      tracker = tracker / 2
    } else {
      tracker = tracker * 3 + 1
    }
    length += 1
  }
  return length
}

func longestChain(under limit: Int) -> Int {
  var longest: Int = 0
  for n in 1..<limit {
    if collatzLength(of: n) > collatzLength(of: longest) {
      longest = n
    }
  }
  return longest
}

print(longestChain(under: 1000000))
